flink-runtime/src/main/java/org/apache/flink/runtime/operators/hash/ReduceHashTable.java

```diff
@@ -0,0 +1,1014 @@
+			compactOrThrow();
+			insert(record);
+			return;
+		}
+
+		numElements++;
+		resizeTableIfNecessary();
+	}
+
+	private void resizeTableIfNecessary() throws IOException {
+		if (enableResize && numElements > numBuckets) {
+			final long newNumBucketSegments = 2L * bucketSegments.length;
+			// Checks:
+			// - we can't handle more than Integer.MAX_VALUE buckets
+			// - don't take more memory than the free memory we have left
+			// - the buckets shouldn't occupy more than half of all our memory
+			if (newNumBucketSegments * numBucketsPerSegment < Integer.MAX_VALUE &&
+				newNumBucketSegments - bucketSegments.length < freeMemorySegments.size() &&
+				newNumBucketSegments < numAllMemorySegments / 2) {
+				// do the resize
+				rebuild(newNumBucketSegments);
+			}
+		}
+	}
+
+	/**
+	 * WARNING: Doing any other operation on the table invalidates the iterator! (Even
+	 * using getMatchFor of a prober!)
+	 */
+	public final class EntryIterator implements MutableObjectIterator<T> {
+
+		private final long endPosition;
+
+		public EntryIterator() {
+			endPosition = recordArea.getAppendPosition();
+			if (endPosition == 0) {
+				return;
+			}
+			recordArea.setReadPosition(0);
+		}
+
+		@Override
+		public T next(T reuse) throws IOException {
+			if (endPosition != 0 && recordArea.getReadPosition() < endPosition) {
+				// Loop until we find a non-abandoned record.
+				// Note: the last record in the record area can't be abandoned.
+				while (true) {
+					final boolean isAbandoned = recordArea.readLong() == ABANDONED_RECORD;
+					reuse = recordArea.readRecord(reuse);
+					if (!isAbandoned) {
+						return reuse;
+					}
+				}
+			} else {
+				return null;
+			}
+		}
+
+		@Override
+		public T next() throws IOException {
+			return next(buildSideSerializer.createInstance());
+		}
+	}
+
+	/**
+	 * Returns an iterator that can be used to iterate over all the elements in the table.
+	 * WARNING: Doing any other operation on the table invalidates the iterator! (Even
+	 * using getMatchFor of a prober!)
+	 * @return the iterator
+     */
+	@Override
+	public EntryIterator getEntryIterator() {
+		return new EntryIterator();
+	}
+
+	/**
+	 * This function reinitializes the bucket segments,
+	 * reads all records from the record segments (sequentially, without using the pointers or the buckets),
+	 * and rebuilds the hash table.
+	 */
+	private void rebuild() throws IOException {
+		rebuild(bucketSegments.length);
+	}
+
+	/** Same as above, but the number of bucket segments of the new table can be specified. */
+	private void rebuild(long newNumBucketSegments) throws IOException {
+		// Get new bucket segments
+		releaseBucketSegments();
+		allocateBucketSegments((int)newNumBucketSegments);
+
+		T record = buildSideSerializer.createInstance();
+		try {
+			EntryIterator iter = getEntryIterator();
+			recordArea.resetAppendPosition();
+			recordArea.setWritePosition(0);
+			while ((record = iter.next(record)) != null) {
```

Comment:
should check for `closed` variable. Same for other longer running method such as `emit`.